physical/raft/fsm.go

```diff
@@ -243,6 +251,90 @@ func writeSnapshotMetaToDB(metadata *raft.SnapshotMeta, db *bolt.DB) error {
 	return nil
 }
 
+func (f *FSM) upgradeConfiguration() error {
+
+	config := f.latestConfig.Load().(*ConfigurationValue)
+
+	// When a fresh cluster is being brought up, there won't be a config. Do
+	// nothing in that case.
+	if config == nil {
+		return nil
+	}
+
+	upgraded := false
+	for _, srv := range config.Servers {
+		if srv.Id == f.localID {
+			// DesiredSuffrage not being set is the upgrade criterion. Use the
+			// current suffrage of the node in the raft cluster as the desired
+			// suffrage, as there is no other better option.
+			if srv.DesiredSuffrage == "" {
+				switch srv.Suffrage {
+				case int32(raft.Nonvoter):
+					srv.DesiredSuffrage = "non-voter"
+				default:
+					srv.DesiredSuffrage = "voter"
+				}
+				upgraded = true
+			}
+			// Bring the intent to the fsm instance. This will be used by
+			// autopilot to keep the non-voters from getting promoted.
+			f.desiredSuffrage = srv.DesiredSuffrage
+		}
+	}
+	if upgraded {
+		if err := f.persistConfig(config); err != nil {
+			return err
+		}
+		f.latestConfig.Store(config)
+	}
+	return nil
+}
+
+// This is called when a node successfully joins the cluster. This intent should
+// land in the stored configuration. If the config isn't available yet, we still
+// go ahead and store the intent in the fsm. During the next updated to the
+// configuration, this intent will be persisted.
+func (f *FSM) witnessSuffrage(desiredSuffrage string) error {
+	config := f.latestConfig.Load().(*ConfigurationValue)
+	if config == nil {
+		// This will be persisted when the fsm receives the `LogConfiguration`
+		// log.
+		// TODO: Do we want to explicitly create the config entry here?
+		f.desiredSuffrage = desiredSuffrage
+		return nil
+	}
+
+	for _, srv := range config.Servers {
+		if srv.Id == f.localID {
+			srv.DesiredSuffrage = desiredSuffrage
+		}
+	}
+
+	if err := f.persistConfig(config); err != nil {
+		return err
+	}
+
+	f.latestConfig.Store(config)
+	f.desiredSuffrage = desiredSuffrage
+
+	return nil
+}
+
+func (f *FSM) persistConfig(config *ConfigurationValue) error {
+	configBytes, err := proto.Marshal(config)
+	if err != nil {
+		return err
+	}
+
+	return f.db.Update(func(tx *bolt.Tx) error {
+		b, err := tx.CreateBucketIfNotExists(configBucketName)
+		if err != nil {
+			return err
+		}
+		return b.Put(latestConfigKey, configBytes)
```

Comment:
Are you sure you want to be writing to the same key as the existing raft config storage?  I'd assumed the idea was to create an entirely new record in the config bucket, rather than augmenting what was already there.

Reply:
I thought, this was the idea. But now that you ask, it maybe is a good idea to have it separate since we don't have to wait on any other log messages to arrive. I'll think a bit more on this.

Reply:
Addressed now.